### ISSUE TITLE
Allow pundit_role to take a block in addition to a symbol

### ISSUE THREAD

**Issue by rgraff** [rmosolgo/graphql-ruby#3039]
**Is your feature request related to a problem? Please describe.**

The required pundit_role on a type may vary based on the record's state.  Currently, you can handle this with custom roles on the policy.

For example:

```
class Type::Post
  pundit_role :author_if_unpublished
end
```

or multiple Types

```
class Type::PublishedPost
  pundit role: nil
end

class Type::UnpublishedPost
  pundit_role :author
end
```

Both solutions are workable until you have more complicated situations and then they become frustrating.

**Describe the solution you'd like**

Allow pundit role to receive a block:

```
class Type::Post
   pundit_role { |r| r.published? ? nil : :author }
```

or allow a pundit_role method in the class

```
class Type::Post
   def pundit_role(record)
     record.published? ? nil : :author
   end
end
```

I'm open to submitting a PR but unsure how to do so given this is a Pro feature.

**Comment by rmosolgo** [comment]
Hey, thanks for the suggestion, I agree that'd be a nice feature. Sorry, there's not a great way to make a PR to graphql-pro, I think your suggestion of adding a class method (like `def self.pundit_role_for(object, user)`) is really promising. Let me take a look and follow up!

**Comment by rmosolgo** [comment]
👋 I just shipped graphql-pro 1.15.4 with a method like this (`pundit_role_for(object, context)`), you can see the updated docs here: https://graphql-ruby.org/authorization/pundit_integration.html#custom-policy-lookup 

If you give it a try, let me know how it goes!

**Comment by rgraff** [comment]
@rmosolgo Yes! Thank you. That works for us.

I'm using it a bit differently then the docs, since we have a lot of variation by type. Maybe update the docs to show how to use on a per class basis without the CustomPolicy?

```ruby
class Type::Post
   def self.pundit_role_for(object, context)
     post.published? ? nil : :author
   end
end
```
